File: leetcode/CPP/1D-Dymanic-Programming/0070-Climbing-Stairs.cpp
/*
    0070. Climbing Stairs   (Easy)

    You are climbing a staircase. It takes n steps to reach the top. 

    Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top? 

    Example 1:

    Input: n = 2
    Output: 2
    Explanation: There are two ways to climb to the top.
    1. 1 step + 1 step
    2. 2 steps
*/

#include <iostream>
#include <vector>

class DynamicProgramming1D {
    public:
        int climbStairs(int n) {
            if (n <= 1) return 1;   // Base case for 1 step
            if (n == 2) return 2;   // Baase case for 2 steps

            int prev1 = 1;      // Ways to reach the step n-1
            int prev2 = 2;      // Ways to reach the step n-2
            int current;

            for (int i = 3; i <= n; i++) {
                current = prev1 + prev2;    // Ways to reach the current step
                prev1 = prev2;              // Update prev1 to prev2
                prev2 = current;            // Update prev2 to current 
            }
            return current;
        }

        int climbStairsDP(int steps) {
            // Base cases: 1 way to climb 0 or 1 step
            if (steps == 0 || steps == 1) {
                return 1;
            }

            // A vector to store the number of ways to climb to each step 
            std::vector<int> waysToClimb(steps + 1);
            waysToClimb[0] = waysToClimb[1] = 1;

            // Compute the number of ways to climb to each step from 2 to steps
            for (int step = 2; step <= steps; ++step) {
                waysToClimb[step] = waysToClimb[step - 1] + waysToClimb[step - 2];
            }

            return waysToClimb[steps];
        }
};

int main() {
    int n = 2;

    DynamicProgramming1D dynamic;
    
    int result = dynamic.climbStairs(n);
    int result_dynamic = dynamic.climbStairsDP(n);

    std::cout << "Climbing Stairs no DP : " << result << std::endl;

    std::cout << "Climbing Stairs with DP : " << result_dynamic << std::endl;

    return 0;
}

File: leetcode/OCaml/1D-Dynamic-Programming-OCaml/0070-Climbing-Stairs.ml
(* 
  0070. Climbing Stairs (Easy)
  
  You are climbing a staircase. It takes n steps to reach the top.

  Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

  Example 1:

  Input: n = 2
  Output: 2
  Explanation: There are two ways to climb to the top.
  1. 1 step + 1 step
  2. 2 steps
*)

let climbingStairs n = 
  if n <= 1 then 1
  else if n = 2 then 2
  else 
    let rec aux i prev1 prev2 = 
      if i > n then prev2
      else 
        let current = prev1 + prev2 in 
        aux (i + 1) prev2 current
      in 
      aux 3 1 2

let climbingStairsDP n = 
    let waysToClimb = Array.make (n + 1) 0 in 
    waysToClimb.(0) <- 1; 
    waysToClimb.(1) <- 1;
    for i = 2  to n do 
      waysToClimb.(i) <- waysToClimb.(i - 1) + waysToClimb.(i - 2)
    done; 
    waysToClimb.(n)

let () = 
    let test n expected = 
      let result = climbingStairs n in 
      Printf.printf "Climb stairs original %d = %d (exptected %d)\n" n result expected
    in 
    test 2 2; 
    test 3 3;
    test 4 5;
    test 5 8


